File: problem2.java
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// removes the N-th node from the end of a linked list.
// It uses a two-pointer approach with a fast pointer and a slow pointer.
public class problem2 {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode start = new ListNode();
        start.next = head;
        ListNode fast = start;
        ListNode slow = start;

        for(int i = 1;i <= n; ++i)
            fast = fast.next;

        while(fast.next!=null){
            fast = fast.next;
            slow = slow.next;
        }

        slow.next = slow.next.next;
        return start.next;
    }
}

File: problem-2.go
/*
time complexity : O(n)
space complexity : O(1)
*/
package Linked_List_1

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummyHead := &ListNode{-1, head}
	slow := dummyHead
	fast := dummyHead
	counter := 0
	for counter < n {
		fast = fast.Next
		counter++
	}
	for fast.Next != nil {
		slow = slow.Next
		fast = fast.Next
	}
	slow.Next = slow.Next.Next
	return dummyHead.Next
}
